gnovm/pkg/gnolang/preprocess.go

```diff
@@ -3366,18 +3366,44 @@ func getResultTypedValues(cx *CallExpr) []TypedValue {
 // NOTE: Generally, conversion happens in a separate step while leaving
 // composite exprs/nodes that contain constant expression nodes (e.g. const
 // exprs in the rhs of AssignStmts).
+//
+// Array-related expressions like `len` and `cap` are manually evaluated
+// as constants, even if the array itself is not a constant. This evaluation
+// is handled independently of the rest of the constant evaluation process,
+// bypassing machine.EvalStatic.
 func evalConst(store Store, last BlockNode, x Expr) *ConstExpr {
 	// TODO: some check or verification for ensuring x
-	// is constant?  From the machine?
-	m := NewMachine(".dontcare", store)
-	m.PreprocessorMode = true
+	var cx *ConstExpr
+	if clx, ok := x.(*CallExpr); ok {
+		t := evalStaticTypeOf(store, last, clx.Args[0])
+		if ar, ok := unwrapPointerType(baseOf(t)).(*ArrayType); ok {
+			fv := clx.Func.(*ConstExpr).V.(*FuncValue)
+			switch {
+			case fv.Name == "cap":
+				fallthrough
+			case fv.Name == "len":
```

Comment:
Very nitpick: It seems you can use a switch statement based on `fv.Name`.
```suggestion
			switch fv.Name {
			case "cap":
				fallthrough
			case "len":
```